Source/Core/Heap.js

```diff
@@ -0,0 +1,194 @@
+/*global define*/
+define([
+        './Check',
+        './defaultValue',
+        './defined',
+        './defineProperties'
+    ], function(
+        Check,
+        defaultValue,
+        defined,
+        defineProperties) {
+    'use strict';
+
+    /**
+     * @alias Heap
+     * @constructor
+     * @private
+     *
+     * @param {Function} comparator The comparator to use for the heap. If comparator(a, b) is less than 0, sort a to a lower index than b, otherwise sort to a higher index.
+     */
+    function Heap(comparator) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('comparator', comparator);
+        //>>includeEnd('debug');
+
+        this._comparator = comparator;
+        this._data = [];
+        this._length = 0;
+        this.maximumSize = 0;
+    }
+
+    defineProperties(Heap.prototype, {
+        /**
+         * Gets the internal data in the heap.
+         *
+         * @memberof Heap.prototype
+         *
+         * @type {Array}
+         * @readonly
+         */
+        data : {
+            get : function() {
+                return this._data;
+            }
+        },
+
+        /**
+         * Gets the length of the heap.
+         *
+         * @memberof Heap.prototype
+         *
+         * @type {Number}
+         * @readonly
+         */
+        length : {
+            get : function() {
+                return this._length;
+            }
+        }
+    });
+
+    function swap(data, a, b) {
+        var temp = data[a];
+        data[a] = data[b];
+        data[b] = temp;
+    }
+
+    /**
+     * Resizes the internal array of the heap.
+     *
+     * @param {Number} [length] The length to resize internal array to. Defaults to the current size of the heap.
+     */
+    Heap.prototype.reserve = function(length) {
+        length = defaultValue(length, this._length);
+        var data = this._data;
+        data.length = length;
+    };
+
+    /**
+     * Heapify. Update the heap so that index and all descendants satisfy the heap property.
+     *
+     * @param {Number} index The starting index to heapify from.
+     */
+    Heap.prototype.heapify = function(index) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.number.greaterThanOrEquals('index', index, 0);
+        //>>includeEnd('debug');
+
+        var length = this._length;
+        var comparator = this._comparator;
+        var data = this._data;
+        var candidate = -1;
+
+        while(1) {
+            var left = 2 * (index + 1) - 1;
```

Comment:
Depending on how this gets optimized, it might be better to write
```
var left = right - 1;
```
or extract ` 2 * (index + 1)` into a local.